File: ios/Capacitor/CapacitorTests/PluginCallAccessorTests.m
#import <XCTest/XCTest.h>
#import <Capacitor/Capacitor.h>
#import <Capacitor/CAPBridgedJSTypes.h>
#import "CapacitorTests-Swift.h"

// interface for this class
@interface PluginCallAccessorTests : XCTestCase
@property (strong, nonatomic) CAPPluginCall* call;
@end

@implementation PluginCallAccessorTests

- (void)setUp {
    [super setUp];
    NSDate* date = [[NSDate alloc] initWithTimeIntervalSinceReferenceDate:632854800];
    NSISO8601DateFormatter *formatter = [[NSISO8601DateFormatter alloc] init];
    NSDictionary* options = @{@"testString":@"foo",
                              @"testDict": @{@"testSubkey":@"sub value"},
                              @"testFloat": @3.14159,
                              @"testDateObject": date,
                              @"testDateString": [formatter stringFromDate:date],
                              @"testBoolTrue": @TRUE,
                              @"testBoolFalse": @FALSE};
    [self setCall:[[CAPPluginCall alloc] initWithCallbackId:@"test" options:options  success:NULL error:NULL]];
}

- (void)testStringAccessor {
    NSString* value = [[self call] getString:@"testString" defaultValue:NULL];
    XCTAssertEqual(value, @"foo");
    
    value = [[self call] getString:@"badString" defaultValue:NULL];
    XCTAssertNil(value);
    
    value = [[self call] getString:@"badString" defaultValue:@"default"];
    XCTAssertEqual(value, @"default");
}

- (void)testDateObjectAccessor {
    NSDate* value = [[self call] getDate:@"testDateObject" defaultValue:NULL];
    XCTAssertEqual([value timeIntervalSinceReferenceDate], 632854800);
    
    value = [[self call] getDate:@"badString" defaultValue:NULL];
    XCTAssertNil(value);

    NSDate *defaultDate = [NSDate date];
    value = [[self call] getDate:@"badString" defaultValue:defaultDate];
    XCTAssertEqual(value, defaultDate);
}

- (void)testDateStringAccessor {
    NSDate* objectValue = [[self call] getDate:@"testDateObject" defaultValue:NULL];
    NSDate* stringValue = [[self call] getDate:@"testDateString" defaultValue:NULL];
    XCTAssertNotNil(objectValue);
    XCTAssertNotNil(stringValue);
    XCTAssertEqual(objectValue, stringValue);
}

- (void)testObjectAccessor {
    NSDictionary* value = [[self call] getObject:@"testDict" defaultValue:NULL];
    XCTAssertEqual([value objectForKey:@"testSubkey"], @"sub value");
    
    value = [[self call] getObject:@"badString" defaultValue:NULL];
    XCTAssertNil(value);
    
    value = [[self call] getObject:@"badString" defaultValue:@{@"defaultKey":@"default"}];
    XCTAssertEqual([value objectForKey:@"defaultKey"], @"default");
}

- (void)testNumberAccessor {
    NSNumber* value = [[self call] getNumber:@"testFloat" defaultValue:NULL];
    XCTAssertNotNil(value);
    XCTAssertTrue([value isEqualToNumber:@3.14159]);
    
    value = [[self call] getNumber:@"badString" defaultValue:NULL];
    XCTAssertNil(value);
    
    value = [[self call] getNumber:@"badString" defaultValue:@100];
    XCTAssertEqual([value intValue], 100);
    
    value = [[self call] getNumber:@"testBoolTrue" defaultValue:NULL];
    XCTAssertNotNil(value);
    XCTAssertEqual([value boolValue], TRUE);
}

- (void)testBoolAccessor {
    BOOL value = [[self call] getBool:@"testBoolTrue" defaultValue:false];
    XCTAssertTrue(value);
    
    value = [[self call] getBool:@"testBoolFalse" defaultValue:true];
    XCTAssertFalse(value);
    
    value = [[self call] getBool:@"badString" defaultValue:true];
    XCTAssertTrue(value);
    
    value = [[self call] getBool:@"badString" defaultValue:false];
    XCTAssertFalse(value);
}
@end

File: ios/Tests/CapacitorTests/PluginCallAccessorTests.swift
import Foundation
import Testing
@testable import Capacitor

struct PluginCallAccessorTests {
    private static let referenceDate = Date(timeIntervalSinceReferenceDate: 632854800)

    private static func makeCall() -> CAPPluginCall {
        let formatter = ISO8601DateFormatter()
        let options: [String: Any] = [
            "testString": "foo",
            "testDict": ["testSubkey": "sub value"],
            "testFloat": 3.14159,
            "testDateObject": referenceDate,
            "testDateString": formatter.string(from: referenceDate),
            "testBoolTrue": true,
            "testBoolFalse": false
        ]
        return CAPPluginCall(callbackId: "test", methodName: "test", options: options, success: { _, _ in }, error: { _ in })
    }

    @Test func stringAccessor() {
        let call = Self.makeCall()
        #expect(call.getString("testString") == "foo")
        #expect(call.getString("badString") == nil)
        #expect(call.getString("badString", defaultValue: "default") == "default")
    }

    @Test func dateObjectAccessor() {
        let call = Self.makeCall()
        #expect(call.getDate("testDateObject")?.timeIntervalSinceReferenceDate == 632854800)
        #expect(call.getDate("badString") == nil)

        let defaultDate = Date()
        #expect(call.getDate("badString", defaultValue: defaultDate) == defaultDate)
    }

    @Test func dateStringAccessor() {
        let call = Self.makeCall()
        let objectValue = call.getDate("testDateObject")
        let stringValue = call.getDate("testDateString")
        #expect(objectValue != nil)
        #expect(stringValue != nil)
        #expect(objectValue == stringValue)
    }

    @Test func objectAccessor() {
        let call = Self.makeCall()
        let value = call.getObject("testDict")
        #expect(value?["testSubkey"] as? String == "sub value")
        #expect(call.getObject("badString") == nil)
    }

    @Test func numberAccessor() {
        let call = Self.makeCall()
        var value = call.getNumber("testFloat")
        #expect(value == NSNumber(value: 3.14159))

        value = call.getNumber("badString")
        #expect(value == nil)

        value = call.getNumber("badString", defaultValue: 100)
        #expect(value?.intValue == 100)

        value = call.getNumber("testBoolTrue")
        #expect(value?.boolValue == true)
    }

    @Test func boolAccessor() {
        let call = Self.makeCall()
        #expect(call.getBool("testBoolTrue", defaultValue: false) == true)
        #expect(call.getBool("testBoolFalse", defaultValue: true) == false)
        #expect(call.getBool("badString", defaultValue: true) == true)
        #expect(call.getBool("badString", defaultValue: false) == false)
    }
}
